File: 0397_integer_replacement/solution.rb
# LeetCode 0397 - Integer Replacement
# https://leetcode.com/problems/integer-replacement/

class Solution
  def integer_replacement(n)
    steps = 0
    while n != 1
      if n.even?
        n /= 2
      elsif n == 3 || n % 4 == 1
        n -= 1
      else
        n += 1
      end
      steps += 1
    end
    steps
  end

  alias_method :integerReplacement, :integer_replacement
end

File: 0397_integer_replacement/solution.c
// LeetCode 0397 - Integer Replacement
// https://leetcode.com/problems/integer-replacement/

int integerReplacement(int n) {
    unsigned long long value = (unsigned long long)n;
    int steps = 0;

    while (value != 1) {
        if (value % 2 == 0) {
            value /= 2;
        } else if (value == 3 || value % 4 == 1) {
            value -= 1;
        } else {
            value += 1;
        }
        steps += 1;
    }

    return steps;
}
